src/axom/core/detail/FlatTable.hpp

```diff
@@ -0,0 +1,334 @@
+    hash ^= hash >> 32;
+    hash *= 0x94d049bb133111ebULL;
+    return hash;
+  }
+};
+
+// We follow the design of boost::unordered_flat_map, which uses a 128-bit chunk
+// of metadata for each group of 15 buckets.
+// This is split up into an "overflow bit", and 15 bytes representing the
+// state of each bucket.
+//
+// Each bucket byte has the value:
+// - 0 if the bucket is "deleted"
+// - 1 to signal the end of the bucket array.
+// - if the bucket has an element, a reduced hash in the range [2, 255].
+//
+// The overflow bit acts as a Bloom filter to determine if probing should
+// terminate.
+struct GroupBucket
+{
+  constexpr static std::uint8_t Empty = 0;
+  constexpr static std::uint8_t Sentinel = 1;
+  constexpr static int InvalidSlot = -1;
+
+  constexpr static int Size = 15;
+
+  GroupBucket() : data {0ULL, 0ULL} { }
+
+  int getEmptyBucket() const
+  {
+    for(int i = 0; i < Size; i++)
+    {
+      if(metadata.buckets[i] == GroupBucket::Empty)
+      {
+        return i;
+      }
+    }
+    // Bucket not found.
+    return InvalidSlot;
+  }
+
+  int nextFilledBucket(int start_index) const
+  {
+    for(int i = start_index + 1; i < Size; i++)
+    {
+      // We intentionally don't check for the sentinel here. This gives us the
+      // index of the sentinel bucket at the end, which is needed as a "stop"
+      // for forward iteration.
+      if(metadata.buckets[i] != GroupBucket::Empty)
+      {
+        return i;
+      }
+    }
+    return InvalidSlot;
+  }
+
+  template <typename Func>
+  int visitHashBucket(std::uint8_t hash, Func&& visitor) const
+  {
+    std::uint8_t reducedHash = reduceHash(hash);
+    for(int i = 0; i < Size; i++)
+    {
+      if(metadata.buckets[i] == reducedHash)
+      {
+        visitor(i);
+      }
+    }
+    return InvalidSlot;
+  }
+
+  void setBucket(int index, std::uint8_t hash)
+  {
+    metadata.buckets[index] = reduceHash(hash);
+  }
+
+  void clearBucket(int index) { metadata.buckets[index] = Empty; }
+
+  void setOverflow(std::uint8_t hash)
+  {
+    std::uint8_t hashOfwBit = 1 << (hash % 8);
+    metadata.ofw |= hashOfwBit;
+  }
+
+  bool getMaybeOverflowed(std::uint8_t hash) const
+  {
+    std::uint8_t hashOfwBit = 1 << (hash % 8);
+    return (metadata.ofw & hashOfwBit);
+  }
+
+  bool hasSentinel() const { return metadata.buckets[Size - 1] == Sentinel; }
+
+  void setSentinel() { metadata.buckets[Size - 1] = Sentinel; }
+
+  // We need to map hashes in the range [0, 255] to [2, 255], since 0 and 1
+  // are taken by the "empty" and "sentinel" values respectively.
+  static std::uint8_t reduceHash(std::uint8_t hash)
+  {
+    return (hash < 2) ? (hash + 8) : hash;
+  }
+
+  union alignas(16)
+  {
+    struct
+    {
+      std::uint8_t ofw;
+      std::uint8_t buckets[Size];
+    } metadata;
+    std::uint64_t data[2];
+    static_assert(
+      sizeof(metadata) == sizeof(data),
+      "FlatMap::SwissTable::Bucket: sizeof(data_bytes) != sizeof(data)");
```

Comment:
FlatMap::SwissTable::Bucket?  I would think flat_map::GroupBucket, or something like that.  (Please check throughout file.)